docs/src/main/sphinx/develop/tests.md

```diff
@@ -0,0 +1,424 @@
+
+The approach to use `Assumptions.abort()` to skip a test, especially deep in the
+call stack, makes it difficult to debug tests failures. The `abort()` works by
+throwing an exception, which can be caught by intervening code inadvertently,
+leading to misleading stack traces and test failures.
+
+### Avoid test inheritance
+
+Inheritance of tests creates unnecessary complexity. Keep tests simple and use
+composition if necessary.
+
+## Avoid helper assertions
+
+The required usage of AssertJ provides a rich set of assertions, that typically
+makes custom helper assertions unnecessary. Custom assertions often make tests
+harder to follow and debug.
+
+If you decide a helper assertion is needed, consider the following details:
+
+* Start the name with `assert`, for example `assertSomeLogicWorks`
+* Prefer private and static
+
+## Examples
+
+The following examples showcase recommended and discourage practices.
+
+### Concurrency for tests
+
+Use `PER_CLASS` for instances because `QueryAssertions` is too expensive to
+create per-method, and a allow parallel execution of tests with `CONCURRENT`:
+
+```java
+@TestInstance(PER_CLASS)
+@Execution(CONCURRENT)
+public class TestJoin
+{
+    private final QueryAssertions assertions = new QueryAssertions();
+
+    @AfterAll
+    public void teardown()
+    {
+        assertions.close();
+    }
+
+    @Test
+    public void testXXX()
+    {
+        assertThat(assertions.query("""
+            ...
+            """))
+            .matches("...");
+    }
+}
+```
+
+### Avoid manual lifecycle management
+
+Avoid managing the lifecycle of a Closeable like a connection with
+`@BeforeEach`/`@AfterEach` to reduce overhead:
+
+```java
+@TestInstance(PER_METHOD)
+class Test
+{
+    private Connection connection;
+
+    @BeforeEach
+    public void setup()
+    {
+        // WRONG: create this in the test method using try-with-resources
+        connection = newConnection();
+    }
+
+    @AfterEach
+    public void teardown()
+    {
+        connection.close();
+    }
+
+    @Test
+    public void test()
+    {
+        ...
+    }
+}
+```
+
+Using a try with resources approach allows clean parallelization of tests and
+includes automatic memory management:
+
+```java
+@TestInstance(PER_METHOD)
+class Test
+{
+
+    @Test
+    public void testSomething()
+    {
+        try (Connection connection = newConnection();) {
+            ...
+        }
+    }
+
+    @Test
+    public void testSomethingElse()
+    {
+        try (Connection connection = newConnection();) {
+            ...
+        }
+    }
+}
+```
+
+## Avoid fake abstractions
+
+Avoid using fake abstraction for tests.
+
+```java
+@DataProvider(name = "data")
+public void test(boolean flag)
+{
+    // WRONG: use separate test methods
+    assertEqual(
+        flag ? ... : ...,
+        flag ? ... : ...);
+}
+```
+
+TBD .. instead do this
+
+```java
+tbd
+```
+
+## Avoid custom parallelization
+
+Do not develop a custom parallel test execution framework:
```

Comment:
the problem this solves is that the test frameworks (neither JUnit not TestNG) don't allow controlling number of concurrent test methods per class.

e.g. how do you for example have a test class which mostly spends time waiting for I/O or some failure and reduce it's runtime?

Reply:
> then you have the different classes run tests in parallel?

Still doesn't solve the problem. The concurrency of test is global in JUnit, ie. if you say 4 concurrent tests you will only have 4 no matter how many test classes you split it into.

You can only override it on a maven module basis via surefire config but that means a separate maven module (or profile) for each such "group of tests".